generator/src/main/java/org/apache/kafka/message/MessageDataGenerator.java

```diff
@@ -398,181 +418,365 @@ private void generateClassReader(String className, StructSpec struct,
+            ifMember(__ -> {
+                buffer.printf("%s = _readable.readUnsignedVarint() - 1;%n", lengthVar);
+            }).
+            ifNotMember(__ -> {
+                if (type.isString()) {
+                    buffer.printf("%s = _readable.readShort();%n", lengthVar);
+                } else if (type.isBytes() || type.isArray()) {
+                    buffer.printf("%s = _readable.readInt();%n", lengthVar);
+                } else {
+                    throw new RuntimeException("Can't handle variable length type " + type);
+                }
+            }).
+            generate(buffer);
+        buffer.printf("if (%s < 0) {%n", lengthVar);
         buffer.incrementIndent();
-        if (generateInverseVersionCheck(parentVersions, struct.versions())) {
+        VersionConditional.forVersions(nullableVersions, possibleVersions).
+            ifNotMember(__ -> {
+                buffer.printf("throw new RuntimeException(\"non-nullable field %s " +
+                    "was serialized as null\");%n", name);
+            }).
+            ifMember(__ -> {
+                buffer.printf("%snull%s", assignmentPrefix, assignmentSuffix);
+            }).
+            generate(buffer);
+        buffer.decrementIndent();
+        if (type.isString()) {
+            buffer.printf("} else if (%s > 0x7fff) {%n", lengthVar);
             buffer.incrementIndent();
-            headerGenerator.addImport(MessageGenerator.UNSUPPORTED_VERSION_EXCEPTION_CLASS);
-            buffer.printf("throw new UnsupportedVersionException(\"Can't read " +
-                "version \" + version + \" of %s\");%n", className);
+            buffer.printf("throw new RuntimeException(\"string field %s " +
+                "had invalid length \" + %s);%n", name, lengthVar);
+            buffer.decrementIndent();
+        }
+        buffer.printf("} else {%n");
+        buffer.incrementIndent();
+        if (type.isString()) {
+            buffer.printf("%s_readable.readString(%s)%s",
+                assignmentPrefix, lengthVar, assignmentSuffix);
+        } else if (type.isBytes()) {
+            buffer.printf("byte[] newBytes = new byte[%s];%n", lengthVar);
+            buffer.printf("_readable.readArray(newBytes);%n");
+            buffer.printf("%snewBytes%s", assignmentPrefix, assignmentSuffix);
+        } else if (type.isArray()) {
+            FieldType.ArrayType arrayType = (FieldType.ArrayType) type;
+            if (isStructArrayWithKeys) {
+                headerGenerator.addImport(MessageGenerator.IMPLICIT_LINKED_HASH_MULTI_COLLECTION_CLASS);
+                buffer.printf("%s newCollection = new %s(%s);%n",
+                    collectionType(arrayType.elementType().toString()),
+                        collectionType(arrayType.elementType().toString()), lengthVar);
+            } else {
+                headerGenerator.addImport(MessageGenerator.ARRAYLIST_CLASS);
+                buffer.printf("ArrayList<%s> newCollection = new ArrayList<%s>(%s);%n",
```

Comment:
I wonder if you have given any thought to limiting allocations like this. For example, in the case of the byte array, we may be able to validate the size using the available bytes in the request

Reply:
I do think we're kind of goofy to allow arrays with 2**31 elements.  There must be a reasonable maximum we could set lower than that.  But there will probably be some compatibility implications to this, so it will take time to impose a reasonable limit now....